File: src/main/java/net/slipp/domain/qna/SnsConnection.java
package net.slipp.domain.qna;

import javax.persistence.Column;
import javax.persistence.Embeddable;
import javax.persistence.EnumType;
import javax.persistence.Enumerated;

import net.slipp.domain.ProviderType;

import org.apache.commons.lang3.StringUtils;

@Embeddable
public class SnsConnection {
    @Enumerated(EnumType.STRING)
    @Column(name = "sns_type", nullable = true, columnDefinition = ProviderType.COLUMN_DEFINITION)
    private ProviderType snsType;

    @Column(name = "post_id", length = 100, nullable = true)
    private String postId;
    
    @Column(name = "group_id", length = 100, nullable = true)
    private String groupId;

    @Column(name = "sns_answer_count", nullable = false)
    private int snsAnswerCount = 0;

    public SnsConnection() {
    }

    public SnsConnection(ProviderType snsType, String postId) {
        this(snsType, postId, null);
    }
    
    public SnsConnection(ProviderType snsType, String postId, String groupId) {
        this.snsType = snsType;
        this.postId = removeId(postId);
        this.groupId = groupId;
    }

    public boolean isConnected() {
        return !StringUtils.isBlank(postId);
    }
    
    public boolean isGroupConnected() {
        return !StringUtils.isBlank(groupId);
    }

    public ProviderType getSnsType() {
        return snsType;
    }

    public String getPostId() {
        return postId;
    }
    
    public String getGroupId() {
		return groupId;
	}

    public int getSnsAnswerCount() {
        return snsAnswerCount;
    }

    public void updateAnswerCount(int answerCount) {
        this.snsAnswerCount = answerCount;
    }
    
    static String removeId(String postId) {
    	if (StringUtils.isBlank(postId)) {
    		return postId;
    	}
    	
    	if (!postId.contains("_")) {
    		return postId;
    	}
    	
    	String[] postIds = postId.split("_");
    	
    	if (postIds.length != 2) {
    		return postId;
    	}
    	
    	return postIds[1];
	}

    @Override
    public int hashCode() {
        final int prime = 31;
        int result = 1;
        result = prime * result + ((postId == null) ? 0 : postId.hashCode());
        result = prime * result + ((snsType == null) ? 0 : snsType.hashCode());
        return result;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj)
            return true;
        if (obj == null)
            return false;
        if (getClass() != obj.getClass())
            return false;
        SnsConnection other = (SnsConnection) obj;
        if (postId == null) {
            if (other.postId != null)
                return false;
        } else if (!postId.equals(other.postId))
            return false;
        if (snsType != other.snsType)
            return false;
        return true;
    }

    @Override
    public String toString() {
        return "SnsConnection [snsType=" + snsType + ", postId=" + postId + ", snsAnswerCount=" + snsAnswerCount + "]";
    }
}

File: src/main/scala/net/slipp/domain/qna/SnsConnection.scala
package net.slipp.domain.qna

import javax.persistence.Column
import javax.persistence.Embeddable
import javax.persistence.EnumType
import javax.persistence.Enumerated

import net.slipp.domain.ProviderType
import net.slipp.domain.user.SocialUser
import net.slipp.support.jpa.DomainModel
import org.apache.commons.lang3.StringUtils

object SnsConnection {
  private[qna] def removeId(postId: String): String = {
    if (StringUtils.isBlank(postId)) {
      return postId
    }
    if (!postId.contains("_")) {
      return postId
    }
    val postIds: Array[String] = postId.split("_")
    if (postIds.length != 2) {
      return postId
    }
    return postIds(1)
  }
}

@Embeddable
class SnsConnection(sType: ProviderType, pId: String, gId: String) extends DomainModel {
  @Enumerated(EnumType.STRING)
  @Column(name = "sns_type", nullable = true, columnDefinition = "enum ('facebook', 'twitter', 'google', 'slipp')")
  private var snsType = sType

  @Column(name = "post_id", length = 100, nullable = true)
  private var postId = pId

  @Column(name = "group_id", length = 100, nullable = true)
  private var groupId = gId

  @Column(name = "sns_answer_count", nullable = false)
  private var snsAnswerCount = 0

  def this(snsType: ProviderType, postId: String) = this(snsType, postId, null)

  def this() = this(null, null, null)

  def isConnected: Boolean = {
    return !StringUtils.isBlank(postId)
  }

  def isGroupConnected: Boolean = {
    return !StringUtils.isBlank(groupId)
  }

  def getSnsType: ProviderType = {
    return snsType
  }

  def getPostId: String = {
    return postId
  }

  def getGroupId: String = {
    return groupId
  }

  def getSnsAnswerCount: Int = {
    return snsAnswerCount
  }

  def updateAnswerCount(answerCount: Int) {
    this.snsAnswerCount = answerCount
  }

  def migrationPostId(providerUserId: String) {
    if (isGroupConnected) {
      return;
    }

    if (postId.contains(providerUserId)) {
      return
    }

    this.postId = providerUserId + "_" + postId
  }
}
